diff --git a/src/minecraft/net/minecraft/src/FontRenderer.java b/src/minecraft/net/minecraft/src/FontRenderer.java
index 0904b44..9c87dd6 100644
--- a/src/minecraft/net/minecraft/src/FontRenderer.java
+++ b/src/minecraft/net/minecraft/src/FontRenderer.java
@@ -17,6 +17,7 @@ public class FontRenderer
 {
     public static boolean betterFontsEnabled = true;
     public StringCache stringCache;
+    public boolean dropShadowEnabled = true;
 
     /**
      * Compiled regular expression pattern for matching color codes in a string
@@ -198,8 +199,10 @@ public class FontRenderer
             {
                 String fontName = config.getFontName("SansSerif");
                 int fontSize = config.getFontSize(18);
+                boolean antiAlias = config.getBoolean("font.antialias", false);
+                dropShadowEnabled = config.getBoolean("font.dropshadow", true);
 
-                stringCache.setDefaultFont(fontName, fontSize);
+                stringCache.setDefaultFont(fontName, fontSize, antiAlias);
                 System.out.println("BetterFonts configuration loaded");
             }
         }
@@ -332,7 +335,11 @@ public class FontRenderer
             par1Str = bidiReorder(par1Str);
         }
 
-        int i = renderString(par1Str, par2 + 1, par3 + 1, par4, true);
+        int i = 0;
+        if(dropShadowEnabled)
+        {
+            i = renderString(par1Str, par2 + 1, par3 + 1, par4, true);
+        }
         i = Math.max(i, renderString(par1Str, par2, par3, par4, false));
         return i;
     }
